core/test/utils/assertions.hpp

```diff
@@ -166,6 +176,78 @@ ::testing::AssertionResult matrices_near_impl(
 }
 
 
+template <typename MatrixData1, typename MatrixData2>
+::testing::AssertionResult matrices_near_sparsity_impl(
+    const std::string &first_expression, const std::string &second_expression,
+    const std::string &tolerance_expression, const MatrixData1 &first,
+    const MatrixData2 &second, double tolerance)
+{
+    auto num_rows = first.size[0];
+    auto num_cols = first.size[1];
+    if (num_rows != second.size[0] || num_cols != second.size[1]) {
+        return ::testing::AssertionFailure()
+               << "Expected matrices of equal size\n\t" << first_expression
+               << " is of size [" << num_rows << " x " << num_cols << "]\n\t"
+               << second_expression << " is of size [" << second.size[0]
+               << " x " << second.size[1] << "]";
+    }
+
+    auto fst_it = begin(first.nonzeros);
+    auto snd_it = begin(second.nonzeros);
+    auto fst_end = end(first.nonzeros);
+    auto snd_end = end(second.nonzeros);
+    using nz_type_f = typename std::decay<decltype(*fst_it)>::type;
+    using nz_type_s = typename std::decay<decltype(*snd_it)>::type;
+    for (size_type row = 0; row < num_rows; ++row) {
+        auto cmp_l_f = [](nz_type_f nz, size_type row) { return nz.row < row; };
+        auto cmp_u_f = [](size_type row, nz_type_f nz) { return row < nz.row; };
+        auto cmp_l_s = [](nz_type_s nz, size_type row) { return nz.row < row; };
+        auto cmp_u_s = [](size_type row, nz_type_s nz) { return row < nz.row; };
+        auto col_eq = [](nz_type_f a, nz_type_s b) {
+            return a.column == b.column;
+        };
+        auto fst_row_begin = std::lower_bound(fst_it, fst_end, row, cmp_l_f);
+        auto snd_row_begin = std::lower_bound(snd_it, snd_end, row, cmp_l_s);
+        auto fst_row_end =
+            std::upper_bound(fst_row_begin, fst_end, row, cmp_u_f);
+        auto snd_row_end =
+            std::upper_bound(snd_row_begin, snd_end, row, cmp_u_s);
+        if (std::distance(fst_row_begin, fst_row_end) !=
```

Comment:
Do you need the first condition here ? Doesnt the second one with `std::equal` supersede this ? Or is it just there for an easy exit ?

Reply:
C++11 is missing an [overload of std::equal (7)](https://en.cppreference.com/w/cpp/algorithm/equal) that takes the beginning and end of both ranges and a comparison operator, thus I need to make sure that we don't move past the last entry of the second range.

Reply:
Ah, okay. Thanks!